My report mention

reckyy opened the pull request.

reckyy committed:
📃 chore: mentionreportsテーブルmigrate

reckyy committed:
⚡️ feat: アソシエーション記述

reckyy committed:
⚡️ feat: mentionreportにバリデーション追加

reckyy committed:
⚡️ feat: 翻訳追加

reckyy committed:
⚡️ feat: 翻訳追加

reckyy committed:
📃 chore: 翻訳ファイル編集

reckyy committed:
⚡️ feat: 日報言及create, update機能作成

reckyy committed:
🐛 fix: updateがうまくいかない不具合を修正

本人が既に言及済みで、言及先を削除、もしくは追加した場合にDBの更新が行われていなかった点を修正

reckyy committed:
🍀 refactor: 共通処理をまとめた

reckyy committed:
💯 test: rubocop

reckyy committed:
💯 test: erblint

reckyy committed:
⚡️ feat: MentionReportテーブルの一意性validation追加

yoshitsugu commented on app/controllers/reports_controller.rb:
```diff
@@ -22,6 +24,7 @@ def create
     @report = current_user.reports.new(report_params)
 
     if @report.save
+      create_mention(@report.id)
```
こういった複数のデータを変更するような場合、トランザクションが必要ないかは常に気にするようにしてください。
トランザクションがわからなければ調べてみてください。

reckyy replied:
コメントありがとうございます！
今のままだと、日報は保存されても言及は保存されない、みたいな事態が発生するようになっていました。申し訳ございません 🙇 
トランザクションを設定し、もしどちらかの保存がうまくいかなかったら、rollbackするようにしました！

[⚡️ feat: Reportのcreate, updateアクションにトランザクション設定](https://github.com/reckyy/fjord-books_app-2023/pull/7/commits/53dc43def93de516af087ef26b5c4e00b8d0b498)
[🐛 fix: save_with_mentionsの引数がnilだった点を修正](https://github.com/reckyy/fjord-books_app-2023/pull/7/commits/a9a84e0f31c4c955cecdb93ea8fb592d89d5b527)

yoshitsugu commented on app/controllers/reports_controller.rb:
```diff
@@ -44,6 +49,19 @@ def destroy
 
   private
 
+  def extract_report_ids(text, report_url: REPORT_URL)
+    text.scan(report_url).flatten.uniq
+  end
+
+  def create_mention(report_id)
+    mentioned_ids = extract_report_ids(@report.content)
+    return if mentioned_ids.empty?
+
+    mentioned_ids.each do |id|
+      MentionReport.create(report_id:, mentioned_report_id: id)
+    end
+  end
```
Controllerは基本的にModelの呼び出しやレンダリングすべきviewの呼び出し、リダイレクトなど、他の処理の起点やHTTPのレスポンスをどうするかを決めるところです。
あまり複雑なロジックを置くべきではありません。
Controllerにどんどんロジックを積み上げてしまったものをFat Controllerと呼び避けるべきパターンの一つとしてよく言われています。
参考: https://bootcamp.fjord.jp/articles/36#fat-controller

このあたりの処理もModel側に置くべきかと思います。

reckyy replied:
controllerには最低限の処理のみを記載するように、以後気をつけます。 🙇 
Reportモデルに言及日報を抜き出すメソッドを移動しました。
[🍀 refactor: reports_controllerのfatcontroller解消](https://github.com/reckyy/fjord-books_app-2023/pull/7/commits/22ad54143eb3a158a6236097cc79e36b50d087d5)

yoshitsugu commented on app/models/report.rb:
```diff
@@ -3,6 +3,10 @@
   has_many :comments, as: :commentable, dependent: :destroy
+  has_many :mention_reports, dependent: :destroy
+  has_many :mentioning_reports, through: :mention_reports, source: :mentioned_report
+  has_many :swapping_id_mention_reports, class_name: 'MentionReport', foreign_key: 'mentioned_report_id', dependent: :destroy, inverse_of: 'mentioned_report'
```
`swapping_id` はかなり実装の詳細に近い話だと思うので、名前づけとしては出すべきではありません。
そもそも関連テーブルが `mention_reports` となっていて `reports` の一種のように見えるのが原因の一旦だと思うので `report_mentions` にしたほうがよいかなと思います。関連テーブルにおいて、 `mention` が主体であり、 `report` はメンションする対象、ですよね。

reckyy replied:
> swapping_id はかなり実装の詳細に近い話だと思うので、名前づけとしては出すべきではありません。

少しだけ抽象度を上げて、`inverse_of`に変更しました。もしこちらでも具体度が高ければ再考しますので、教えていただければ幸いです。
また、何故具体的な名前づけがいけないのかも併せて教えていただけますと幸いです。
[🐛 fix: モデル名の変更に伴い、has_many関連も修正](https://github.com/reckyy/fjord-books_app-2023/pull/7/commits/d4017d172341b1a3d5869dfc6fb1c4357e9202d1)

> そもそも関連テーブルが mention_reports となっていて reports の一種のように見えるのが原因の一旦だと思うので report_mentions にしたほうがよいかなと思います。関連テーブルにおいて、 mention が主体であり、 report はメンションする対象、ですよね。

おっしゃる通りです。`belongs_to`だからReportに属するMention、なのにモデル名やテーブル名を間違えていました。。申し訳ございません。
モデル名やテーブル名など該当箇所は修正しました。

[🐛 fix: モデル名、テーブル名をReportMentionに変更](https://github.com/reckyy/fjord-books_app-2023/pull/7/commits/e479d767d567933f46e36873b0e290e9fb6f7eb9)

yoshitsugu reviewed (commented):
コメントしました。

reckyy committed:
🐛 fix: モデル名、テーブル名をReportMentionに変更

reckyy committed:
🐛 fix: モデル名の変更に伴い、has_many関連も修正

reckyy committed:
🐛 fix: モデル名の変更に伴い、i18n変更

reckyy committed:
🍀 refactor: reports_controllerのfatcontroller解消

Report言及のロジックをモデルに切り出した

reckyy committed:
⚡️ feat: Reportのcreate, updateアクションにトランザクション設定

reckyy committed:
🐛 fix: save_with_mentionsの引数がnilだった点を修正

引数に@report.idを渡していたが、それが設定されるのはsave後のためnilになっていた。なので、内部でインスタンスのidを参照するように修正。

reckyy committed:
📃 chore: 使用しないテストファイルを削除

yoshitsugu commented on app/models/report.rb:
```diff
@@ -14,4 +18,28 @@ def editable?(target_user)
+
+  def save_with_mentions
+    Report.transaction do
+      (save && save_mentions(id)) || raise(ActiveRecord::Rollback)
```
たしかにこれでもうまくいきますが、ちょっとトリッキーなので、

```ruby
succeeded = save
succeeded = save_mentions(id) if succeeded
raise ActiveRecord::Rollback unless succeeded
```

とかのほうが見やすいかなと思います。

reckyy replied:
順を追って記述した方がこの場合とても見やすいですね、、ありがとうございます。
その通りに修正しました。
[🍀 refactor: createとupdateでのメンション保存メソッドを共通化](https://github.com/reckyy/fjord-books_app-2023/pull/7/commits/81bab0143ee3531774069aef31f0f53ba9318b5a)

yoshitsugu commented on app/models/report.rb:
```diff
@@ -14,4 +18,28 @@ def editable?(target_user)
+
+  def update_with_mentions(report_params, report_id)
+    Report.transaction do
+      (update(report_params) && save_mentions(report_id)) || raise(ActiveRecord::Rollback)
```
同上

また、 `assign_attributes` であらかじめ保存なしで属性を更新しておけば `update` と `save` を分ける必要がなくなるかなと思います。
https://api.rubyonrails.org/classes/ActiveModel/AttributeAssignment.html#method-i-assign_attributes

reckyy replied:
`assign_attributes`、知らなかったです。教えていただきありがとうございます 🙇 
`update`アクションの前にあらかじめ`assign_attributes`で属性を更新し、create, updateともにsaveで保存できるように変更しました。
[🍀 refactor: createとupdateでのメンション保存メソッドを共通化](https://github.com/reckyy/fjord-books_app-2023/pull/7/commits/81bab0143ee3531774069aef31f0f53ba9318b5a)

yoshitsugu commented on app/models/report.rb:
```diff
@@ -14,4 +18,28 @@ def editable?(target_user)
+    mentioned_ids = content.to_s.scan(REPORT_URL).flatten.uniq
+    return true if mentioned_ids.empty?
+
+    mentioned_ids.each do |id|
+      ReportMention.create(report_id:, mentioned_report_id: id)
+    end
```
これだと `each` の返り値が `save_mentions` の返り値になってしまうので、ちゃんと判定できないように思います。

```ruby
> [1,2,3].each { 10 }
=> [1, 2, 3]
```
みたいな感じで、 `each` の返り値は `each` を呼び出したオブジェクトになります。

reckyy replied:
確かに判定できていませんでした。申し訳ございません。
`save_mentions`の返り値をtrueに変更しました。
[🐛 fix: save_mentionsメソッドの返り値をtrueに設定](https://github.com/reckyy/fjord-books_app-2023/pull/7/commits/f3f82fbc39e900b5e6ffd267b84e395dc879d55d)

yoshitsugu replied:
これだと結局 `each` の中身が `false` になったとしても `true` を返すので、意味ないのではと思うのですが、どうでしょうか。

yoshitsugu commented on app/models/report.rb:
```diff
@@ -3,6 +3,10 @@
   has_many :comments, as: :commentable, dependent: :destroy
+  has_many :report_mentions, dependent: :destroy
+  has_many :mentioning_reports, through: :report_mentions, source: :mentioned_report
+  has_many :inverse_of_report_mentions, class_name: 'ReportMention', foreign_key: 'mentioned_report_id', dependent: :destroy, inverse_of: 'mentioned_report'
```
`report_mentioner` (言及する側)と `report_mentionee` (言及される側)でどうでしょうか。まだ意味的に抽象的で対応がとれてそうに思います。

reckyy replied:
[🐛 fix: 参照先を記述し忘れ](https://github.com/reckyy/fjord-books_app-2023/pull/7/commits/c468171b22f5a2f576135887840943b67910e73c)

yoshitsugu reviewed (commented):
コメントしました。

reckyy committed:
🐛 fix: save_mentionsメソッドの返り値をtrueに設定

最後のeachの結果が返り値になっていたため、trueを返せていなかった。

reckyy committed:
🐛 fix: 参照先を記述し忘れ

reckyy committed:
🍀 refactor: createとupdateでのメンション保存メソッドを共通化

save_with_mentionsとupdate_with_mentionsにメソッドを分けていたが（report_paramsの代入タイミングが異なるため）、assign_attributesでupdateの前に一時保存し、両方saveで保存できるようにした

reckyy committed:
💯 test: rubocop

reckyy commented:
> コメントしました。

ありがとうございます！各ご指摘内容について修正いたしましたので、ご確認のほどよろしくお願いいたします。

また、前回のモデルのアソシエーションのご指摘の際に、質問させていただいたのですがこちらに再掲させていただきます。
ご確認いただけますと幸いです。

> swapping_id はかなり実装の詳細に近い話だと思うので、名前づけとしては出すべきではありません。

> 少しだけ抽象度を上げて、inverse_ofに変更しました。もしこちらでも具体度が高ければ再考しますので、教えていただければ幸いです。
また、何故具体的な名前づけがいけないのかも併せて教えていただけますと幸いです。

https://github.com/reckyy/fjord-books_app-2023/pull/7#discussion_r1387416571

yoshitsugu reviewed (commented):
> また、何故具体的な名前づけがいけないのかも併せて教えていただけますと幸いです。

こちらに答えられていなかったですね。
たとえば、

```ruby
def x_plus_one(x)
  x + 1
end
```

を見てどう思います？それよりは


```ruby
def increment(x)
  x + 1
end
```

のほうがやりたいこと(意図)がわかりやくないでしょうか。
また、

```ruby
MON_TUE_WED_THU_FRI = ["月", "火", "水", "木", "金"]
```

より

```ruby
WEEKDAY_NAMES = ["月", "火", "水", "木", "金"]
```

のほうがこの定数の存在意義がわかりやすくないでしょうか。
メソッド名にしても変数名にしても、名前づけというのは一種の抽象化です。
そのものが「何を目的としているのか」「どういう性質があるのか」ということが大事で「中身がどうなっているのか」は重要ではありません。
あまりいい例が思いつかなかったので、かなり作為的な例ですが、

```ruby
def x_minus_one_plus_one_plus_one(x)
    x - 1 + 1 + 1
end
```

というコードがあったとします(絶対ないと思いますが)
で、この作者が後に「あ、普通に+1だけすればいいじゃん」と気づいたとします。
すると

```ruby
def x_minus_one_plus_one_plus_one
  x + 1
end
```

になり、すぐに実態と名前が乖離してしまいます。 
一方 `increment` にしておけば、意味が抽象化されているため、名前は変えなくてすみますよね。
つまり

- 変更時にインターフェース(メソッド名や引数など)を変える必要が発生しづらくなり、メンテナンスしやすくなる
- 読むときにコードの詳細を知らなくてもよいので、コードリーディングがしやすくなる

といったメリットがあります。
これは名前と実装の[結合度](https://ja.wikipedia.org/wiki/%E7%B5%90%E5%90%88%E5%BA%A6)を下げる、と捉えることもできます。

reckyy commented:
とんでもございません。ご返信ありがとうございます！

> メソッド名にしても変数名にしても、名前づけというのは一種の抽象化です。 そのものが「何を目的としているのか」「どういう性質があるのか」ということが大事で「中身がどうなっているのか」は重要ではありません。

とても丁寧に解説していただきありがとうございます。
メソッド名や変数名を命名する時は、中身ではなく本質を考えて行うようにします！
ありがとうございました😭

reckyy committed:
🐛 fix: save_mentionsの戻り値を修正

処理の内容に関わらずいつでもtrueを返すようになっていたので、失敗時にはfalseを返すようにsucceededフラグを設定

reckyy commented:
なぜかhttps://github.com/reckyy/fjord-books_app-2023/pull/7#discussion_r1390504745 に返信できなかったので、こちらに記載させていただきます。

> これだと結局 each の中身が false になったとしても true を返すので、意味ないのではと思うのですが、どうでしょうか。

おっしゃる通りです。安直な修正をしてしまい、申し訳ございません。`save_with_mentions`メソッドと同様に`succeeded`フラグを用意し、処理失敗時には`false`を返すように修正しました。



[🐛 fix: save_mentionsの戻り値を修正](https://github.com/reckyy/fjord-books_app-2023/pull/7/commits/a13d1b57f85f3cd66fcef9fd732baa3c20502c7f)

yoshitsugu commented on app/models/report.rb:
```diff
@@ -14,4 +18,28 @@ def editable?(target_user)
+    mentioned_ids = content.to_s.scan(REPORT_URL).flatten.uniq
+    return succeeded if mentioned_ids.empty?
+
+    mentioned_ids.each do |id|
+      succeeded = false unless ReportMention.create(report_id:, mentioned_report_id: id)
+    end
+    succeeded
```
`any?` や `all?` などを使うと簡単にかけそうです。
https://docs.ruby-lang.org/ja/latest/method/Enumerable/i/any=3f.html
https://docs.ruby-lang.org/ja/latest/method/Enumerable/i/all=3f.html

reckyy replied:
succeededフラグを用いることで、処理が冗長になってしまっていました。（元々`empty?`メソッドがboolean型を返すので、検討すべき事案でした。申し訳ございません）
処理の失敗時に、booleanの値（true, false）を返すように修正いたしました。
[🐛 fix: save_mentions内部で処理失敗時、直接真偽値を返すように変更](https://github.com/reckyy/fjord-books_app-2023/pull/7/commits/f37469153a3db3bb386857675390038f9df978b9)

yoshitsugu reviewed (commented):
細かいところ一点だけコメントしました。

reckyy committed:
🐛 fix: save_mentions内部で処理失敗時、直接真偽値を返すように変更

succeededフラグを削除し、真偽値が戻り値のメソッドを採用

reckyy committed:
💯 test: rubocop

yoshitsugu reviewed (approved):
OKです〜
